accepted/2024/vulnerabilities-in-packages.config-restore.md

```diff
@@ -0,0 +1,93 @@
+# ***Vulnerabilities in packages.config restore***
+
+- [Nikolche Kolev](https://github.com/nkolev92)
+- GitHub Issue <https://github.com/NuGet/Home/issues/12307>
+
+## Summary
+
+This is a continuation of the [vulnerabilities in PackageReference work](../2022/vulnerabilities-in-restore.md), extending matching functionality for packages.config restore.
+
+## Motivation
+
+See [vulnerabilities in PackageReference restore motivation](../2022/vulnerabilities-in-restore.md#motivation).
+
+## Explanation
+
+### Functional explanation
+
+Packages.config restore combines all packages regardless of project. Instead of raising warnings and errors on a per project level, at restore time, they're raised on a per package level.
+
+The configuration knobs for packages.config restore auditing functionality will be NuGet.config based.
+In particular, there will be 2 new configuration keys:
+
+| Key | Acceptable values | Description | Default |
+|-----|-------------------|-------------|---------|
+| auditForPackagesConfig | enable, disable | Enables or disables NuGet Audit for packages config projects | If not specified, the default will be `enable` |
+| auditLevelForPackagesConfig | Critical, high, moderate, low | Configurations the default audit level for NuGet audit for packages config projects |  If not specified, the default will be `disable` |
+
+The audit functionality for packages.config restore will be enabled by default.
+To disable it, one can specify a property in the config section of the configuration file.
+
+```xml
+<configuration>
+    <config>
+        <add key="audit" value="enable" />
+        <add key="auditLevel" value="low" />
+    </config>
+</configuration>
+```
+
+Whenever a vulnerability for a package is discovered, a warning will be raised indicating the severity and advisory url.
+Note that warnings as errors and no warn are not support in packages.config projects, so these warnings will not fail the build.
+
+> Package 'Contoso.Service.APIs' 1.0.3 has a known critical severity vulnerability, https://github.com/advisories/GHSA-1234-5678-9012.
+
+For performance considerations, vulnerability checks are only going to be performed when the packages.config restore downloads a package.
```

Comment:
I don't agree with this.

Firstly, it makes packages.config package audit on dev machines of very limited usefulness. Therefore, my interpretation is that it appears that this is being designed as a CI-only feature. But since packages.config doesn't support warning as errors, it obligates customers to check their CI logs frequently.

Secondly, consider just Azure Pipelines for a moment. Microsoft hosted agents are stateless (wiped clean after every job). However, self-hosted agents are not. Azure Virtual Machine Scale Set agents are configurable stateful or stateless. On stateful machines, the `checkout` task has a parameter to clean, or not, the repo. Therefore, customers who set this to skip cleaning will have populated solution packages folders before restore runs. Furthermore, whther you're on a stateful or stateless CI agent, some customers use [the cache task](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/cache-v2?view=azure-pipelines) to populate the global packages, or solution packages directory, more quickly than NuGet can download and extract packages.

So, even on CI, there are many scenarios where NuGet won't be downloading or extracting a package on restore.

IMO there are too many scenarios where "audit only when downloading packages" will prevent recent advisories from being presented to customers, so I don't think it's a good design.

Reply:
We need a better approach than every time for packages.config. 

We've already gotten some feedback where people have ended up disabling audit for PR because it goes online even when all the packages are on disk. I'm struggling to find that issue now, but i'll figure it out. 

Another alternative is when the audit data changes, but that's a lot more complicated *without* the fast up to date checks for audit data. 

A different idea is only when a package is downloaded *and on first* restore. 
Either way, I think every single time might lead to some noise.

Reply:
Here's one I know of: https://github.com/NuGet/Home/issues/13118#issuecomment-1882409588

Reply:
I changed it to default, and added details in unresolved questions to figure out if we need to be more selective in order not to regress the overall experience.